### ISSUE TITLE
Need to remove tweets that donot belong to python programming language

### ISSUE THREAD

**Issue by yogithapolavarapu** [yogithapolavarapu/Retweet-Bot#1]
Script retweets all the tweets with python in them, It will be appreciated if you could filter it so that tweets about snakes and monty python won't be retweeted.

**Comment by inwardgarden** [comment]
Do the current conditions/phrases not work though- they still get through and are re-tweeted?

For instance I see that you're already checking if the tweet contains "Monty":
`TWITTER_SEARCH_PHRASE_2 = "Monty";`

`&& !tweet.includes(TWITTER_SEARCH_PHRASE_2)`

**Comment by yogithapolavarapu** [comment]
I'm trying to fetch the tweets that are not sensitive or contain the words like Monty to retweet them.
If we place an OR instead of an AND, that would mean any tweet which is sensitive or has a word like Monty will be fetched

for a non sensitive tweet and tweet with no Monty word in it.
!tweet.possibly_sensitive && !tweet.includes("Monty")
!(FALSE) && !(FALSE) -> this will be fetched

for a non sensitive tweet and tweet with monty
!tweet.possibly_sensitive && !tweet.includes("Monty")
!(FALSE) && !(TRUE) -> this will not be fetched

for a non sensitive tweet and tweet with monty with OR instead of AND
!tweet.possibly_sensitive && !tweet.includes("Monty")
!(FALSE) || !(TRUE) -> this will be fetched and we don't want that to happen.

There might be a way to turn it around and make it work, you are welcome to look into it and make a pull request once Hacktoberfest starts

**Comment by inwardgarden** [comment]
Right. I was mistaken in my initial post (that I have now edited) due to misreading it initially. It makes sense to keep the &&. I was thinking about it in reverse.

Does the current code not work at all however? Does it still retweet tweets with "Monty"? None of the recent retweets seem to contain this phrase.

**Comment by yogithapolavarapu** [comment]
Yeah, sadly, they are being retweeted. The rate at which the tweets with monty being retweeted, have reduced but a few tweets creep in from time to time

**Comment by inwardgarden** [comment]
Okay, I'm looking into it. I think it may have to do with the "includes" function, since the tweet object itself is not a string, and I think includes() only works on strings. It is possible that tweet.text.includes("Monty") would work. Though includes() is also case-sensitive, so that may be something to consider too and could explain why it's working partially (maybe?)

I did notice it's even grabbing tweets that don't contain "python", but from users whose usernames contain python. That may be a whole separate issue, though.

**Comment by yogithapolavarapu** [comment]
yes, it will be a seperate issue, once the one with immediate concern is done

**Comment by inwardgarden** [comment]
I have my own test bot running and so far with some changes, it seems to be working. I'm also slightly switching the way it checks for phrases, if that's okay. Essentially instead of growing the && conditions in the if statement to an overwhelming amount, I have set up a function that loops through an array of "excluded words" and checks if they're in the tweet. If any single word is found, the tweet_processor argument will skip the tweet. Still running it periodically to make sure it's actually filtering things. But so far so good, I'll submit in October 👍

**Comment by yogithapolavarapu** [comment]
Sure, thank you for taking time for this

**Comment by inwardgarden** [comment]
Think I found something weird / interesting:
https://github.com/bscottx/Retweet-Bot/issues/1

My test bot account retweeted a Computer Science tweet, but I can't find any mention/reference to "python" anywhere in the tweet, so I have no idea how it happened. It's actually a semi-relevant tweet, but I'm curious why the bot thinks it contains Python... lol 🤔 🤔 🤔

I suppose it has to do with Twitter's querying. Maybe it returns "related" tweets and not just tweets that technically contain the phrase.

**Comment by yogithapolavarapu** [comment]
would using tweet.text instead of tweet solve the issue?

**Comment by inwardgarden** [comment]
I'm not sure. It depends on if the API is still willing to give us "related" results once it's done with any of the directly matching results. I have a feeling the API might always be a little "too" helpful. 

It would limit the relevance query to the tweet text only. Which means it wouldn't check links/photos/media. I guess that could help narrow it down, but it also might hurt the results/quality of the retweets a bit, such as if someone just posts a link to a Python resource, it would overlook that.

I found a couple of additional issues with my filter code. Stuff was still getting through, but that should be fixed now and seems to be working right. I've had to add all sorts of words to the filter because apparently the word "python" is referring to some political stuff now, types of leather, and all sorts of things. The false positives are much less in quantity now.

The test bot is live at https://twitter.com/pythybot if you want to see it.

Basically getting a blast of tweets from fetch_tweets and then running them through very specific criteria seems to be the best way to get a good set of results without limiting the "search" too much.

It's really open-ended. In the `isTweetRelevant` function all sorts of criteria could be added, like requiring that certain parameters contain the word python, like tweet.text (or however you access usernames, etc)

**Comment by inwardgarden** [comment]
I believe I might have discovered why the false positives happen.

It's possible that it's retweeting a retweet of a tweet. So, if anyone in the retweet chain has "python" in their bio, it will probably get looped in.

I don't know if this still works, but Twitter does have (poorly documented) ways of filtering search queries. ` -filter:nativeretweets` and also ` -filter:retweets` in the query should prevent it from retweeting retweets (and only retweet them from the source). Going to test these out. Will keep updating the bot until it feels like the results are good and then will submit the PR

**Comment by inwardgarden** [comment]
I've noticed pretty much all of the good tweets contain the word **python** somewhere, so I think I am going to add that requirement to the filter.

**Comment by yogithapolavarapu** [comment]
Okay., make a pull request when you are done. :smile:

**Comment by inwardgarden** [comment]
I submitted the PR but have made some additional changes since, just FYI (searching the whole tweet object for unwanted words, instead of just the text via `JSON.stringify(tweetobj)`) 

If I need to do a new PR later, that is fine, or if you want the additional changes and git has a way for you to request them, feel free. 😎
